pkg/cloudprovider/aws.go

```diff
@@ -1,32 +1,342 @@
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/wait"
+	utilnet "k8s.io/utils/net"
 )
 
 const (
 	aws = "aws"
 )
 
+var NoNetworkInterfaceError error = errors.New("no retrievable network interface")
+
 // AWS implements the API wrapper for talking to the AWS cloud API
 type AWS struct {
 	CloudProvider
+	region string
+	client *ec2.EC2
 }
 
 func (a *AWS) initCredentials() error {
+	accessKey, err := a.readSecretData("aws_access_key_id")
+	if err != nil {
+		return err
+	}
+	secretKey, err := a.readSecretData("aws_secret_access_key")
+	if err != nil {
+		return err
+	}
+	mySession := session.Must(session.NewSession())
+	a.client = ec2.New(mySession, awsapi.NewConfig().WithCredentials(awscredentials.NewStaticCredentials(accessKey, secretKey, "")).WithRegion(a.region))
 	return nil
 }
 
+// AssignPrivateIP assigns the IP address to the node by re-providing all
+// existing ones + the new one. It does this on a per-IP-family basis (since the
+// AWS API is separated per family). If the IP is already existing: it returns an
+// AlreadyExistingIPError.
 func (a *AWS) AssignPrivateIP(ip net.IP, node *corev1.Node) error {
+	instance, err := a.getInstance(node)
+	if err != nil {
+		return err
+	}
+	networkInterfaces, err := a.getNetworkInterfaces(instance)
+	if err != nil {
+		return err
+	}
+	// Perform the operation against the first interface listed following the
+	// order AWS specifies.
+	for _, networkInterface := range networkInterfaces {
+		addIP := ip.String()
+		keepIPs := []*string{}
+		if utilnet.IsIPv6(ip) {
+			for _, assignedIPv6 := range networkInterface.Ipv6Addresses {
+				if assignedIP := net.ParseIP(*assignedIPv6.Ipv6Address); assignedIP != nil && assignedIP.Equal(ip) {
+					return AlreadyExistingIPError
+				}
+				keepIPs = append(keepIPs, assignedIPv6.Ipv6Address)
+			}
+			keepIPs = append(keepIPs, &addIP)
+			input := ec2.AssignIpv6AddressesInput{
+				NetworkInterfaceId: networkInterface.NetworkInterfaceId,
+				Ipv6Addresses:      keepIPs,
+			}
+			_, err = a.client.AssignIpv6Addresses(&input)
+			if err != nil {
+				return err
+			}
+			return a.waitForCompletion(node, awsapi.StringValueSlice(keepIPs), false)
+		}
+		for _, assignedIPv4 := range networkInterface.PrivateIpAddresses {
+			if assignedIP := net.ParseIP(*assignedIPv4.PrivateIpAddress); assignedIP != nil && assignedIP.Equal(ip) {
+				return AlreadyExistingIPError
+			}
+			keepIPs = append(keepIPs, assignedIPv4.PrivateIpAddress)
+		}
+		keepIPs = append(keepIPs, &addIP)
+		inputV4 := ec2.AssignPrivateIpAddressesInput{
+			NetworkInterfaceId: networkInterface.NetworkInterfaceId,
+			PrivateIpAddresses: keepIPs,
+		}
+		_, err = a.client.AssignPrivateIpAddresses(&inputV4)
+		if err != nil {
+			return err
+		}
+		return a.waitForCompletion(node, awsapi.StringValueSlice(keepIPs), false)
+	}
 	return nil
 }
 
+// ReleasePrivateIP un-assigns the IP address from the node by re-assigning the
+// existing IP addresses - the one to be removed.
```

Comment:
This is no longer accurate, right? You appear to be using a delete function.

Reply:
Indeed it's not. I have changed the doc bloc.